renderdoc/driver/metal/metal_device.cpp

```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+MTL::Library *WrappedMTLDevice::newLibraryWithSource(NS::String *source,
+                                                     MTL::CompileOptions *options, NS::Error **error)
+{
+  MTL::Library *realMTLLibrary;
+  SERIALISE_TIME_CALL(realMTLLibrary =
+                          Unwrap<MTL::Device *>(this)->newLibrary(source, options, error));
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
+  ResourceId id = GetResourceManager()->WrapResource(realMTLLibrary, wrappedMTLLibrary);
+  if(IsCaptureMode(m_State))
+  {
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLDevice_newLibraryWithSource);
+      Serialise_mtlDevice_newLibraryWithSource(ser, this, wrappedMTLLibrary, source, options);
+      chunk = scope.Get();
+    }
+    MetalResourceRecord *record = GetResourceManager()->AddResourceRecord(wrappedMTLLibrary);
+    record->AddChunk(chunk);
+    GetResourceManager()->MarkResourceFrameReferenced(id, eFrameRef_Read);
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+template <typename SerialiserType>
+bool WrappedMTLDevice::Serialise_mtlDevice_newDefaultLibrary(SerialiserType &ser,
+                                                             WrappedMTLDevice *device,
+                                                             WrappedMTLLibrary *library,
+                                                             const void *pData, uint32_t bytesCount)
+{
+  SERIALISE_ELEMENT_LOCAL(Library, GetResID(library)).TypedAs("MTLLibrary"_lit);
+  SERIALISE_ELEMENT(bytesCount);
+  SERIALISE_ELEMENT_ARRAY(pData, bytesCount);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  if(IsReplayingAndReading())
+  {
+    // TODO: implement RD MTL replay
+  }
+  return true;
+}
+
+template <typename SerialiserType>
+bool WrappedMTLDevice::Serialise_mtlDevice_newLibraryWithSource(SerialiserType &ser,
+                                                                WrappedMTLDevice *device,
+                                                                WrappedMTLLibrary *library,
+                                                                NS::String *source,
+                                                                MTL::CompileOptions *options)
+{
+  SERIALISE_ELEMENT_LOCAL(Library, GetResID(library)).TypedAs("MTLLibrary"_lit);
+  SERIALISE_ELEMENT(source);
+  // TODO:SERIALISE_ELEMENT(options);
+
+  SERIALISE_CHECK_READ_ERRORS();
+
+  if(IsReplayingAndReading())
+  {
+    // TODO: implement RD MTL replay
+  }
+  return true;
+}
+
+MTL::Library *WrappedMTLDevice::CreateMTLLibrary(const void *pData, uint32_t bytesCount)
```

Comment:
Is this intended as a wrapper for `newLibrary`? It should share the same name then and (as closely as possible) the same function signature. This does seem like a bit of an interop challenge though.

Reply:
It is a helper method for `Serialise_newDefaultLibrary` and is not required for anything else.
I will put the code directly in the method.
It comes from before metal-cpp where this was implemented in the objective c helper file.

Reply:
1f9a217577fe36c8bc218da0bbb2f3e721f22011